### DISCUSSION TITLE
🚀 Distro 2024.02.0 Released

### DISCUSSION THREAD

**Discussion by Nicogene** [robotology/community#690] [Releases]
# 2024.02.0 Distro Announcement

## Brief description of the announcement

Hello @robotology/everyone  👋

The [iCub project](https://icub.iit.it/) is happy to announce our **Software Distro 2024.02.0** 📦

This is a major release that comprises **binary packages** as well. Therefore, follow the [instructions][1] available in our documentation to upgrade your system and/or download the new dependencies.

You can find the tags composing this distro in our [Software Versioning Table][2].

## Detailed context

This distro is shipped with several updates and fixes, including the ones listed hereafter:

#### 🔘 New joint coupling's sw architecture

We recently re-design how in our software we want to handle the coupling of joints in underactuated systems.
In particular, this new architecture is based on the YARP plugin system, this allows to load runtime the coupling handler, without introducing compile-time dependencies.
First of all we designed a new interface [`yarp::dev::IJointCoupling`](https://github.com/robotology/yarp/blob/master/src/libYARP_dev/src/yarp/dev/IJointCoupling.h) in order to fit the architecture explained as follow.

- https://github.com/robotology/yarp/pull/3026#issue-1909013157

Then we started using this new infrastructure in simulation, in particular the coupling handler of hand mk5 has been already refactored in order to be a plugin:

- https://github.com/icub-tech-iit/ergocub-software/pull/178
- https://github.com/robotology/gazebo-yarp-plugins/pull/671

#### 🔘 `xcub-moveit2` is out now

[`xcub-moveit2`](https://github.com/icub-tech-iit/xcub-moveit2) is a collector of ros2 packages to control in cartesian space both iCub and ergoCub robots by means of `ros2_control` and `MoveIt` tools. 

Check out the project on the [MoveIt website](https://moveit.ros.org/robots/).

#### 🔘 2FOC firmware improvements
Fix a synchronization problem in the quadrature encoder counter reset by index. Now the motors can run smoothly:
- https://github.com/robotology/icub-firmware/pull/455
- https://github.com/robotology/icub-firmware-build/pull/128

#### 🔘 Motor temperature reading
If a motor is equipped with temperature sensor, now it is possibile to get the motor temperature by the API of IMotor interface. See [icub-tech documentation](https://icub-tech-iit.github.io/documentation/temperature_sensors/).   

#### 🔘 IMU test
We added to [`icub-tests`](https://github.com/robotology/icub-tests) a new test for the IMU sensors. The aim is to check the Euler angles values measured by the sensor and compare them to the expected values computed with the forward kinematics parsing the URDF of the robot:

- https://github.com/robotology/icub-tests/pull/65

The data from the encoders and the sensors are acquired and processed *realtime*, and the parameters to run the test could be set in a specific configuration file:

- https://github.com/robotology/icub-tests/pull/67

#### 🔘 `icub-firmware-*` fully integrated in `robotology-superbuild`

Now the `robotology-superbuild` integrates also the `icub-firmware-build` repository.
In a nutshell, this means that since 2024.02.0 release, now if you enable the `ROBOTOLOGY_ENABLE_ICUB_HEAD` 
option in the superbuild, on the clean build of the superbuild you can find the exact version of the firmware that you need to flash on the robot in the `<superbuild_build_directory>/install/share/icub-firmware/` folder.

Furthermore, now the `icub-firmware` binaries are also available in the `icub-firmware` conda package available in the `robotology` channel. For example, if you want to create a `conda` environment called `xcubfirmware` to install the `icub-firmware` corresponding to the distro `2024.02.0`, you can with the following command:
~~~bash
mamba create -n xcubfirmware -c conda-forge -c robotology icub-firmware "robotology-distro==2024.02.0"
~~~

After installation, you  will find the `icub-firmware` files in `$CONDA_PREFIX\share\icub-firmware` in Linux/macOS and `%CONDA_PREFIX%\Library\share\icub-firmware` on Windows.

For more details, see the corresponding PRs:
  - https://github.com/robotology/icub-firmware-build/pull/124
  - https://github.com/robotology/robotology-superbuild/pull/1559
  - https://github.com/robotology/ycm/pull/440
  - https://github.com/robotology/robotology-superbuild/pull/1560

#### 🔘 Several important fixes to the wrist-mk2 of ergoCub
- https://github.com/robotology/icub-firmware-models/pull/82
- https://github.com/robotology/icub-firmware/pull/467
- https://github.com/robotology/icub-firmware/pull/469

#### 🔘 Community contributions
- https://github.com/robotology/icub-main/pull/932 from @rustlluk 🎉 

#### 🔘 Replace `embObjInertial` w/ `embObjIMU`

`embObjInertial` was the "legacy" device that handled the inertial sensors on iCub. It was used for old boards (e.g. mtb board) and it was based on an old firmware service that is not used anymore.

We managed to add the support to those old boards in the "modern" `embObjIMU` device and this allowed also to remove both the old device and the old firmware inertial service that is not used anymore, reducing the code duplication and the size of the firmware binaries.

- https://github.com/robotology/robots-configuration/issues/420
- https://github.com/robotology/robots-configuration/issues/595

[1]: https://icub-tech-iit.github.io/documentation/sw_installation/
[2]: https://icub-tech-iit.github.io/documentation/sw_versioning_table/#table

**Comment by traversaro** [comment]
Hello everyone, sorry I forgot to update the release notes in time, but I would like to provide more details on the "`icub-firmware-*` fully integrated in `robotology-superbuild`" functionality.

In a nutshell, this means that since 2024.02.0 release, now if you enable the `ROBOTOLOGY_ENABLE_ICUB_HEAD` 
option in the superbuild, on the clean build of the superbuild you can find the exact version of the firmware that you need to flash on the robot in the `<superbuild_build_directory>/install/share/icub-firmware/` folder.

Furthermore, now the `icub-firmware` binaries are also available in the [`icub-firmware` conda package](https://anaconda.org/robotology/icub-firmware) available in the `robotology` channel. For example, if you want to create a `conda` environment called `xcubfirmware` to install the `icub-firmware` corresponding to the distro `2024.02.0`, you can with the following command:
~~~bash
mamba create -n xcubfirmware -c conda-forge -c robotology icub-firmware "robotology-distro==2024.02.0"
~~~

After installation, you  will find the `icub-firmware` files in `$CONDA_PREFIX\share\icub-firmware` in Linux/macOS or `%CONDA_PREFIX%\Library\share\icub-firmware` on Windows.


fyi @robotology/iit-artificial-mechanical-intelligence @robotology/icub-tech-support-team
